### ISSUE TITLE
Why does the dataloader of DAVIS need to randomly sample images from a video sequence?

### ISSUE THREAD

**Issue by MaxChanger** [xmlyqing00/AFB-URR#4]
Hello, thanks for sharing such a complete code.

Maybe due to my inadequate understanding, I have some questions about the train process, I hope to get your answers.
In the dataloader `DAVIS_Train_DS` of training, each time, **6 randomly shuffled images are sampled** from this video sequence, as mentioned in the paper _[4.2 Main training on the benchmark datasets]_. But in `DAVIS_Test_DS`, the images are read in order.

I want to know, what is the purpose of random sampling? Are there any obvious advantages to this approach?
If I read five or six consecutive images randomly, e.g. [random_id : random_id + 6], will there be any negative effects?

Thanks.

https://github.com/xmlyqing00/AFB-URR/blob/8db9e4bffffc5683da1e0f7d7f0fc0d306909072/dataset/DAVIS_DS.py#L40-L51

**Comment by xmlyqing00** [comment]
Thank you for pointing this out. 
> what is the purpose of random sampling? 

This is a kind of data augmentation. We want to synthesize more different sequences to train our model because the number of videos in the DAVIS dataset is too small.

> Are there any obvious advantages to this approach?

This is a good question, we didn't justify the benefits of random sampling because it is widely used in many other works. I believe this data augmentation could make the model robust.

> If I read five or six consecutive images randomly, e.g. [random_id : random_id + 6], will there be any negative effects?

I didn't try this. It is a good try.
Btw, I guess you are confused about **6 randomly shuffled images are sampled**. We use the first frame to build the feature bank, then the rest five frames are treated as a batch for matching. The length of the randomly generated sequence is only 2. I think using longer sequences could benefit the model but make the training process slower.

As for the codes. I got your point. I read the filenames of the images first. Then I shuffled the filenames and picked the first 6 items. After that, I read the actual images.

**Comment by MaxChanger** [comment]
Thank you for your prompt reply, and make me understand better.
In the test/eval phase, the images are read sequentially, and each image will be segmented and memorized. 
But during the training process, five (unordered) images are treated as a batch at the same time. Will they update the memory bank one by one?  I am a little curious.
In the code, five non-contiguous images are directly sent to the segment function:
https://github.com/xmlyqing00/AFB-URR/blob/8db9e4bffffc5683da1e0f7d7f0fc0d306909072/train.py#L64-L73

Another question is what's mean about ` The length of the randomly generated sequence is only 2`, which sequence?
Thanks.

**Comment by xmlyqing00** [comment]
> In the test/eval phase, the images are read sequentially, and each image will be segmented and memorized.

This is correct.

> But during the training process, five (unordered) images are treated as a batch at the same time. Will they update the memory bank one by one? I am a little curious.

There are B images (B = 6 in our settings) in a batch, the first image is memorized to build the feature bank, the rest images are used for segmentation and back-propagate the loss. Obviously, the lengths of the randomly generated sequences are all 2 (one for memorization, the other one for segmentation). We build and train such 5 sequences simultaneously. 

Since the feature bank is a deterministic module, we don't do the append and update operations during training.

**Comment by MaxChanger** [comment]
Thank you @xmlyqing00 very much for your patient and detailed answers. 👍👍
